pkg/models/vulnerability.go

```diff
@@ -167,3 +174,155 @@ func (v Vulnerability) MarshalYAML() (interface{}, error) {
 
 	return raw, nil
 }
+
+// NormalizedName ensures that the package name is normalized based on ecosystem
+// in accordance to the Vulnerability specification.
+//
+// This is required because currently both GitHub and Pip seem to be a bit
+// inconsistent in their package name handling, so we normalize them
+// to be on the safe side.
+//
+// In the future, it's hoped that this can be improved.
+func (p Package) NormalizedName() string {
+	if p.Ecosystem != lockfile.PipEcosystem {
+		return p.Name
+	}
+
+	// per https://www.python.org/dev/peps/pep-0503/#normalized-names
+	name := regexp.MustCompile(`[-_.]+`).ReplaceAllString(p.Name, "-")
+
+	return strings.ToLower(name)
+}
+
+func (ar Range) containsVersion(pkg lockfile.PackageDetails) bool {
+	if ar.Type != RangeEcosystem && ar.Type != RangeSemVer {
+		return false
+	}
+	// todo: we should probably warn here
+	if len(ar.Events) == 0 {
+		return false
+	}
+
+	vp := semantic.MustParse(pkg.Version, pkg.CompareAs)
+
+	var affected bool
+	for _, e := range ar.Events {
+		if affected {
+			if e.Fixed != "" {
+				affected = vp.CompareStr(e.Fixed) < 0
+			} else if e.LastAffected != "" {
+				affected = e.LastAffected == pkg.Version || vp.CompareStr(e.LastAffected) <= 0
+			}
+		} else if e.Introduced != "" {
+			affected = e.Introduced == "0" || vp.CompareStr(e.Introduced) >= 0
+		}
+	}
+
+	return affected
+}
+
+type Affects []Range
+
+// affectsVersion checks if the given version is within the range
+// specified by the events of any "Ecosystem" or "Semver" type ranges
+func (a Affects) affectsVersion(pkg lockfile.PackageDetails) bool {
+	for _, r := range a {
+		if r.Type != RangeEcosystem && r.Type != RangeSemVer {
```

Comment:
nit: This check seems to be duplicated in `containsVersion`. (This does allow a early exit here though, so remove it from containsVersion?)

Reply:
tbh I might actually just axe this function entirely, and that'll let me get rid of `Affects` too

Reply:
actually I think probably implement `MarshalJSON` for this like I have for `Versions` in which case I'll need to keep the type still 🤔